examples/gno.land/p/demo/int256/int256.gno

```diff
@@ -1,64 +1,87 @@
-// This package provides a 256-bit signed integer type, Int, and associated functions.
 package int256
 
 import (
+	"errors"
+
 	"gno.land/p/demo/uint256"
 )
 
-var one = uint256.NewUint(1)
+var (
+	int1  = NewInt(1)
+	uint0 = uint256.NewUint(0)
+	uint1 = uint256.NewUint(1)
+)
 
 type Int struct {
-	abs *uint256.Uint
-	neg bool
+	value uint256.Uint
 }
 
-// Zero returns a new Int set to 0.
-func Zero() *Int {
-	return NewInt(0)
+// New creates and returns a new Int initialized to zero.
+func New() *Int {
+	return &Int{}
 }
 
-// One returns a new Int set to 1.
-func One() *Int {
-	return NewInt(1)
+// NewInt allocates and returns a new Int set to the value of the provided int64.
+func NewInt(x int64) *Int {
+	return New().SetInt64(x)
 }
 
-// Sign returns:
+// Zero returns a new Int initialized to 0.
 //
-//	-1 if x <  0
-//	 0 if x == 0
-//	+1 if x >  0
-func (z *Int) Sign() int {
-	z.initiateAbs()
+// This function is useful for creating a starting point for calculations or
+// when an explicit zero value is needed.
+func Zero() *Int { return &Int{} }
 
-	if z.abs.IsZero() {
-		return 0
-	}
-	if z.neg {
-		return -1
-	}
-	return 1
-}
-
-// New returns a new Int set to 0.
-func New() *Int {
+// One returns a new Int initialized to one.
+//
+// This function is convenient for operations that require a unit value,
+// such as incrementing or serving as an identity element in multiplication.
+func One() *Int {
 	return &Int{
-		abs: new(uint256.Uint),
+		value: *uint256.NewUint(1),
 	}
 }
 
-// NewInt allocates and returns a new Int set to x.
-func NewInt(x int64) *Int {
-	return New().SetInt64(x)
+// Sign determines the sign of the Int.
+//
+// It returns -1 for negative numbers, 0 for zero, and +1 for positive numbers.
+func (z *Int) Sign() int {
+	if z.IsZero() {
+		return 0
+	}
+	// Right shift the value by 255 bits to check the sign bit.
+	// In two's complement representation, the most significant bit (MSB) is the sign bit.
+	// If the MSB is 0, the number is positive; if it is 1, the number is negative.
+	//
+	// Example:
+	// Original value:  1 0 1 0 ... 0 1  (256 bits)
+	// After Rsh 255:   0 0 0 0 ... 0 1  (1 bit)
+	//
+	// This approach is highly efficient as it avoids the need for comparisons
+	// or arithmetic operations on the full 256-bit number. Instead it reduces
+	// the problem to checking a single bit.
+	//
+	// Additionally, this method will work correctly for all values,
+	// including the minimum possible negative number (which in two's complement
+	// doesn't have a positive counterpart in the same bit range).
+	var temp uint256.Uint
+	if temp.Rsh(&z.value, 255).IsZero() {
```

Comment:
can we just check against arr[3] so it's not necessary to shift the whole for 255? just curious the impact on the performance.

Reply:
I used this approach because `Uint`'s internal array is private. directly manipulating the array was definitely faster than current version, but this method was still sufficiently usable for alternative.